web-api/src/storage/elastic/filter.rs

```diff
@@ -249,6 +357,110 @@ enum Clause<'a> {
     MustNot(MustNot<'a>),
 }
 
+fn fully_merge_range(mut clause: Vec<Clause<'_>>) -> Vec<Clause<'_>> {
+    fn is_range<'a>(clause: &Clause<'a>) -> Option<&'a DocumentPropertyId> {
+        if let Clause::Range(range) = clause {
+            Some(range.field)
+        } else {
+            None
+        }
+    }
+
+    let mut i = 0;
+    while i < clause.len() {
```

Comment:
as said before, this will panic because the condition in the for loop is only checked once at loop initialization. as we remove elements the vec shrinks and we will get out of bounds panic. that's why we need to increment the counter manually and check the condition at every iteration.